flow/diff_context.cc

```diff
@@ -0,0 +1,167 @@
+                         double frame_device_pixel_ratio,
+                         PaintRegionMap& this_frame_paint_region_map,
+                         const PaintRegionMap& last_frame_paint_region_map)
+    : rects_(std::make_shared<std::vector<SkRect>>()),
+      frame_size_(frame_size),
+      frame_device_pixel_ratio_(frame_device_pixel_ratio),
+      this_frame_paint_region_map_(this_frame_paint_region_map),
+      last_frame_paint_region_map_(last_frame_paint_region_map) {}
+
+void DiffContext::BeginSubtree() {
+  state_stack_.push_back(state_);
+  state_.rect_index_ = rects_->size();
+}
+
+void DiffContext::EndSubtree() {
+  FML_DCHECK(!state_stack_.empty());
+  state_ = std::move(state_stack_.back());
+  state_stack_.pop_back();
+}
+
+DiffContext::State::State()
+    : dirty(false), cull_rect(kGiantRect), rect_index_(0) {}
+
+void DiffContext::PushTransform(const SkMatrix& transform) {
+  state_.transform.preConcat(transform);
+  SkMatrix inverse_transform;
+  // Perspective projections don't produce rectangles that are useful for
+  // culling for some reason.
+  if (!transform.hasPerspective() && transform.invert(&inverse_transform)) {
+    inverse_transform.mapRect(&state_.cull_rect);
+  } else {
+    state_.cull_rect = kGiantRect;
+  }
+}
+
+Damage DiffContext::GetDamage(const SkIRect& accumulated_buffer_damage) const {
+  SkRect framebuffer = SkRect::Make(accumulated_buffer_damage);
+  framebuffer.join(damage_);
+  SkRect net(damage_);
+
+  for (const auto& r : readbacks_) {
+    if (r.rect.intersects(net)) {
+      net.join(r.rect);
+    }
+    if (r.rect.intersects(framebuffer)) {
+      framebuffer.join(r.rect);
+    }
+  }
+
+  Damage res;
+  framebuffer.roundOut(&res.buffer_damage);
+  net.roundOut(&res.surface_damage);
+
+  SkIRect frame_clip = SkIRect::MakeSize(frame_size_);
+  res.buffer_damage.intersect(frame_clip);
+  res.surface_damage.intersect(frame_clip);
+  return res;
+}
+
+bool DiffContext::PushCullRect(const SkRect& clip) {
+  return state_.cull_rect.intersect(clip);
+}
+
+void DiffContext::MarkSubtreeDirty(const PaintRegion& previous_paint_region) {
+  FML_DCHECK(!IsSubtreeDirty());
+  if (previous_paint_region.is_valid()) {
+    AddDamage(previous_paint_region);
+  }
+  state_.dirty = true;
+}
+
+void DiffContext::AddPaintRegion(const SkRect& rect) {
+  SkRect r(rect);
+  if (r.intersect(state_.cull_rect)) {
+    state_.transform.mapRect(&r);
+    if (!r.isEmpty()) {
+      rects_->push_back(r);
+      if (IsSubtreeDirty()) {
+        AddDamage(r);
+      }
+    }
+  }
+}
+
+void DiffContext::AddExistingPaintRegion(const PaintRegion& region) {
+  // Adding paint region for retained layer implies that current subtree is not
+  // dirty
+  FML_DCHECK(!IsSubtreeDirty());
+  FML_DCHECK(region.is_valid());
+  rects_->insert(rects_->end(), region.begin(), region.end());
+}
+
+void DiffContext::AddReadbackRegion(const SkRect& rect) {
+  SkRect r(rect);
+  state_.transform.mapRect(&r);
```

Comment:
As mention before, right now filter bounds are calculated in local coordinates. I can see though how this might cause subtle glitches around the edges with downscaled BackdropFilterLayer. I'll give this another look.

Reply:
Resolved with https://github.com/flutter/engine/pull/21824/commits/72d0b75167f771af74eeb8f70cdf563a581cbb4a.